.github/workflows/ci.yml

```diff
@@ -118,75 +144,20 @@ jobs:
     # coverage-collecting jobs to this workflow but didn't update this, that's
     # why.
-    needs:
-      - "coverage"
-    runs-on: "ubuntu-latest"
-    steps:
-      - name: "Check out Tahoe-LAFS sources"
-        uses: "actions/checkout@v2"
+    needs: coverage
+    runs-on: ubuntu-latest
+    container: python:3-slim
 
-      - name: "Finish Coveralls Reporting"
+    steps:
+      - name: Indicate completion to coveralls.io
         run: |
-          # coveralls-python does have a `--finish` option but it doesn't seem
-          # to work, at least for us.
-          # https://github.com/coveralls-clients/coveralls-python/issues/248
-          #
-          # But all it does is this simple POST so we can just send it
-          # ourselves.  The only hard part is guessing what the POST
-          # parameters mean.  And I've done that for you already.
-          #
-          # Since the build is done I'm going to guess that "done" is a fine
-          # value for status.
-          #
-          # That leaves "build_num".  The coveralls documentation gives some
-          # hints about it.  It suggests using $CIRCLE_WORKFLOW_ID if your job
-          # is on CircleCI.  CircleCI documentation says this about
-          # CIRCLE_WORKFLOW_ID:
-          #
-          # Observation of the coveralls.io web interface, logs from the
-          # coveralls command in action, and experimentation suggests the
-          # value for PRs is something more like:
-          #
-          #    <GIT MERGE COMMIT HASH>-PR-<PR NUM>
-          #
-          # For branches, it's just the git branch tip hash.
-
-          # For pull requests, refs/pull/<PR NUM>/merge was just checked out
-          # by so HEAD will refer to the right revision.  For branches, HEAD
-          # is also the tip of the branch.
-          REV=$(git rev-parse HEAD)
-
-          # We can get the PR number from the "context".
-          #
-          # https://docs.github.com/en/free-pro-team@latest/developers/webhooks-and-events/webhook-events-and-payloads#pull_request
-          #
-          # (via <https://github.community/t/github-ref-is-inconsistent/17728/3>).
-          #
-          # If this is a pull request, `github.event` is a `pull_request`
-          # structure which has `number` right in it.
-          #
-          # If this is a push, `github.event` is a `push` instead but we only
-          # need the revision to construct the build_num.
-
-          PR=${{ github.event.number }}
-
-          if [ "${PR}" = "" ]; then
-              BUILD_NUM=$REV
-          else
-              BUILD_NUM=$REV-PR-$PR
-          fi
-          REPO_NAME=$GITHUB_REPOSITORY
-
-          curl \
-            -k \
-            https://coveralls.io/webhook?repo_token=$COVERALLS_REPO_TOKEN \
-            -d \
-            "payload[build_num]=$BUILD_NUM&payload[status]=done&payload[repo_name]=$REPO_NAME"
+          pip3 install --upgrade coveralls
```

Comment:
This hunk increases overall complexity of this step a lot, though it does so by getting rid of a little local complexity and replacing it with a lot of foreign complexity.  I'm not sure this is a great trade-off for the CI system - at least as long as the CI system itself (our configuration for it and the third-party systems) is untested (apart from incidentally as part of normal development workflow for the rest of Tahoe).

Also, all this stuff that's being deleted was here for a reason.  I guess that reason is obsolete now?  I see that the upstream issue is closed, anyway.  If it's really better to use coveralls-python than curl here then it would be good to resolve that independently of the other change here.  That will make it easier to observe the consequences of this change separately from the rest of the changes in this branch.

Reply:
I documented the need for consistently using one version of coveralls-python, here:

https://github.com/tahoe-lafs/tahoe-lafs/blob/c6f404f4a22e556b90a9ca58075931e9e63bcc38/.github/workflows/ci.yml#L104-L114

Adding Python 3 would add coveralls-python 3.0.1.  Mixing two versions of coveralls-python and then curl led to some of the reports hanging around as incomplete.

Also, we weren't signaling end-of-tests using curl exactly in the recommended manner.  The `payload` should have conformed to this structure:

```json
{
  "payload": {
       "build_num": 1234,
       "status": "done"
     }
}
```
We had also added a `payload[repo_name]`.  I could not be sure if it affected anything, but, after switching to coveralls-python 3.0.1 in all the steps, coveralls.io marked all jobs as complete.